File: year-2/programming-design/java/MotorcycleContestClientServer/MotorcycleServer/src/main/java/networking/AbstractConcurrentServer.java
package networking;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.net.Socket;

public abstract class AbstractConcurrentServer extends AbstractServer {
    private static final Logger LOGGER = LogManager.getLogger(AbstractConcurrentServer.class.getName());

    public AbstractConcurrentServer(int port) {
        super(port);
    }

    public AbstractConcurrentServer() {
    }

    protected void processRequest(Socket client) {
        LOGGER.info("processing request");
        Thread workerThread = createWorker(client);
        LOGGER.info("starting processing request thread");
        workerThread.start();
    }

    protected abstract Thread createWorker(Socket client);
}

File: year2/mpp/c#/MotorcycleClientServer/MotorcycleServer/Networking/AbstractConcurrentServer.cs
using System.Net.Sockets;
using System.Threading;

namespace MotorcycleServer.Networking
{
    public abstract class AbstractConcurrentServer : AbstractServer
    {
        public AbstractConcurrentServer(string host, int port) : base(host, port) { }

        public override void ProcessRequest(TcpClient client)
        {
            // Logger.Info("processing client request");
            Thread thread = CreateWorker(client);
            thread.Start();
        }

        protected abstract Thread CreateWorker(TcpClient client);
    }
}